File: ProblemSolving/450DSA/JavaScript/src/string/AllPermutations.js
/**
 * @author Anirudh Sharma
 * 
 * Given a string S. The task is to print all permutations of a given string.
 * 
 * Constraints:
 * 1 ≤ T ≤ 10
 * 1 ≤ size of string ≤ 5
 */
const findPermutations = (s) => {
    // Array to store result
    const permutations = [];
    // Recursive call
    findPermutationsHelper(s, "", permutations);
    return permutations;
};

const findPermutationsHelper = (s, prefix, permutations) => {
    // Length of the string
    const n = s.length;
    // Special case
    if (n === 0) {
        permutations.push(prefix);
        return;
    }
    for (let i = 0; i < n; i++) {
        findPermutationsHelper(s.substring(0, i) + s.substring(i + 1, n), prefix + s.charAt(i), permutations);
    }
};

const main = () => {
    console.log(findPermutations("ABC"));
    console.log(findPermutations("ABSG"));
};

main();

File: ProblemSolving/450DSA/Python/src/string/AllPermutations.py
"""
@author Anirudh Sharma

Given a string S. The task is to print all permutations of a given string.

Constraints:
1 ≤ T ≤ 10
1 ≤ size of string ≤ 5
"""


def findPermutationsHelper(s, prefix, permutations):
    # Length of the string
    n = len(s)
    # Special case
    if n == 0:
        permutations.append(prefix)
        return
    for i in range(n):
        findPermutationsHelper(s[0:i] + s[i+1:n], prefix + s[i], permutations)


def findPermutations(s):
    # list to store the result
    permutations = []
    # Recursive call
    findPermutationsHelper(s, "", permutations)
    return permutations


if __name__ == "__main__":
    print(findPermutations("ABC"))
    print(findPermutations("ABSG"))